Bart as single-library + introducing Python module library pyBART v2

Takishima opened the pull request:
New take on #131 but separating the changes in multiple commits for easier review.

uecker commented:
Excellent. Thank you! I will slowly add some comments.

uecker commented on src/misc/debug.h:
```diff
@@ -19,8 +21,20 @@ extern int debug_level;
 
 enum debug_levels { DP_ERROR, DP_WARN, DP_INFO, DP_DEBUG1, DP_DEBUG2, DP_DEBUG3, DP_DEBUG4, DP_TRACE, DP_ALL };
-extern void debug_printf(int level, const char* fmt, ...);
-extern void debug_vprintf(int level, const char* fmt, va_list ap);
```
Please preserve these functions as is. The default debug_printf_trace could then just use these.

Takishima replied:
Ok.

Takishima replied:
Ok will do.

uecker commented on src/misc/debug.h:
```diff
@@ -19,8 +21,20 @@ extern int debug_level;
+			       int level, const char* fmt, ...);
+
+// To get the proper function name, file and line when spoofing UTrace
+#define debug_printf(level, ...)					\
```
These redefinition should be optional depending on a preprocessor variable

Takishima replied:
Agreed.

Takishima replied:
Good idea.

uecker replied:
The same for debug_print_dims. Thank you!

uecker commented on src/misc/debug.h:
```diff
@@ -9,6 +10,7 @@
 
 #include <stdarg.h>
 #include <stddef.h>
+#include <stdbool.h>
```
This include seems unncessary.

Takishima replied:
Ok.

uecker commented on src/misc/debug.c:
```diff
@@ -97,7 +97,15 @@ void debug_vprintf_trace(const char* func_name,
+		// take care of the trailing newline often present...
+		if (fmt_copy[strlen(fmt)-1] == '\n') {
+			fmt_copy[strlen(fmt)-1] = '\0';
+		}
```
This should then go into the new debug_printf_trace function and could directly call into the vendor wrappers. Or why do we need this output wrappers?

Takishima replied:
Done.

uecker commented on src/misc/misc.h:
```diff
@@ -10,6 +11,7 @@
 
 #include <stdlib.h>
 #include <stddef.h>
+#include <stdbool.h>
```
include not needed.

Takishima replied:
Ok.

uecker commented on src/misc/misc.h:
```diff
@@ -52,12 +54,24 @@ extern void warn_nonnull_ptr(void*);
 #ifndef __cplusplus
 extern noreturn void error(const char* str, ...);
+#elif defined __clang__
+extern _Noreturn void error(const char* str, ...);
```
What for? The standard is noreturn (which expands to _Noreturn=

Takishima replied:
I had some compilation problems which have since disappeared...

Takishima commented:
I have pushed some changes addressing some of the comments.

Regarding the indentation issues with the second commit, these only occur because the indentation is correct for the final state after the last commit.
In order to split the changes into multiple commits, I simply selectively committed chunks of changes together (ie. without properly re-indenting the changes).

uecker commented on src/misc/misc.h:
```diff
@@ -67,8 +67,19 @@ extern float quickselect(float *arr, unsigned int n, unsigned int k);
+extern int mini_cmdline_bool(int* argcp, char* argv[], char flag_char, int expected_args, const char* usage_str, const char* help_str, _Bool* flag);
+
+#define mini_cmdline(acp, av, e_args, u_str, h_str)			\
+     do {								\
```
I would rather avoid these macros and the change in API. How about we 'error' instead of 'exit' and the original versions of 'mini_cmdline' and 'cmdline'. We could then use 'longjmp' excactly as in io_error. What do you think?

Takishima replied:
Since these functions are only used in very few specific locations and are not a central part of the BART framework, I don't think changing the API is all that bad.
Plus, from a design perspective, I don't like functions which job should be only limited with parsing arguments exiting the program on their own. I feel it should be the caller's responsibility to deal with any error and then exit if need be.

The effects on client code should also be fairly minimal, as only custom commands written outside of BART calling `mini_cmdline_bool` directly would be impacted by those changes (and in those cases the code would simply not compile, avoiding running with unexpected behaviour). A comment in the code above the declaration of `mini_cmdline_bool` could then instruct users on how they can fix their code.

uecker replied:
There is no sensible way to recover from an incorrect command-line. Why do you think it is good from a design perspective to add error handling?

But even worse, you do not even add this error handling, you just re-implement  the same API using macros ;-)

To add insult to injury, it is not even the same API: Instead of calling exit, the macro returns from the calling function. This is really unusual and surprising behavior and now makes it impossible to use these functions anywhere else than in main.

Takishima replied:
I agree with you that recovering from an invalid command line is impossible and that a program should not try to continue afterwards, but instead indicate to the user that an error occurred and terminate. Calling the `exit`function is one way of doing so which works well for standalone programs like BART.

However, in programming languages that have proper exception handling, failure to parse command line arguments is usually handled by throwing an exception that gives the possibility of each caller function to act upon it and do some cleanup if necessary (such as closing file descriptors, freeing heap allocated memory, etc.). Additionally, throughout this process normal stack unwinding takes place to make sure that no memory accidentally becomes unaccounted for. Unfortunately, C has no such mechanism and `longjmp` emulates only part of those functionalities, which is why I tried to avoid relying on it as much as possible.

I also did not know that these functions were intended to be called elsewhere in BART, which make my modifications indeed inappropriate.

Anyway, in light of your comments, I will revert those changes and revert to calling the `error` function which will then make the program jump back to the BART main function (via `longjmp` if enabled) and subsequently return and exit with an error code.

uecker replied:
Well, there are not necessarily called elsewhere in BART, but there is a lot of software build on top of BART in my lab or reusing parts of BART (and maybe elsewhere). But I tend agree with your analysis, and I agree that the longjmp solution is also not ideal. Maybe the best choice would be to return the error and change all the callers to check it, but this seems overkill at this time.

uecker commented on src/ecalib.c:
```diff
@@ -46,6 +46,8 @@ int main_ecalib(int argc, char* argv[])
 	long calsize[3] = { 24, 24, 24 }; 
 	int maps = 2;
+	int debug_level_save = debug_level;
+	int debug_level_local = debug_level;
```
Could we save and restore 'debug_level' in 'main_bart' ? You are using this is main entry function, or?

Takishima replied:
Done.

uecker commented:
I will start cherry-picking and revising some individual commits. I you want, you can rebase. But if it is too much trouble, don't worry you can also just continue to work on your branch and leave the merging to me.

Takishima commented:
Ok, so I've rebased it and addressed the `debug_level` issue inside the BART main function.

uecker commented on src/misc/debug.c:
```diff
@@ -141,7 +141,54 @@ void debug_vprintf_trace(const char* func_name,
 {
+#ifdef USE_LOG_BACKEND
+	char fmt_copy[1024] = "";
+	strcpy(fmt_copy, fmt);
```
could this not just be applied to tmp below?

Takishima replied:
Yes, you're right, I've missed that for some reason.

uecker commented on src/misc/debug.c:
```diff
@@ -141,7 +141,54 @@ void debug_vprintf_trace(const char* func_name,
+
+	switch(level) {
+	case DP_ERR:
+#  ifdef USE_LOG_SIEMENS_BACKEND
```
ok, how about a single 'vendor_log' function which takes a 'level' argument? And all different backends could simply export the same function (as they are mutually exclusive anyway). This would remove this boilerplate code.

Takishima replied:
Sounds good, I will make the necessary changes.

uecker commented on src/misc/debug.h:
```diff
@@ -39,13 +43,80 @@ extern void debug_vprintf_trace(const char* func_name,
+#  define EXTERN_FUNC extern
+#endif /* __cplusplus */
+
+#ifdef BART_SUPER_DEBUG
```
we already have debug_level to control the amount of debug output.

Takishima replied:
I will remove this and replace it with calls to `debug_printf`

uecker commented on src/misc/debug.h:
```diff
@@ -39,13 +43,80 @@ extern void debug_vprintf_trace(const char* func_name,
 #define TRACE()	debug_trace("%s:%d %s\n", __FILE__, __LINE__, __func__)
 
 
+#ifdef USE_LOG_BACKEND
```
this should go in the header of the corresponding c/cc file. (and simplied to one function)

Takishima replied:
I prefer not to add an extra header file for just a single function, so I've left the prototype of `vendor_log`in debug.h

uecker commented on src/misc/mmiocc.cc:
```diff
@@ -0,0 +1,617 @@
+ * a BSD-style license which can be found in the LICENSE file.
+ *
+ * Authors:
+ * 2017-2018 Nguyen Damien <damien.nguyen@alumni.epfl.ch>
```
I would find it a bit more useful and fitting when this were written C, but ok.
I assume you want to be able to wrap existing memory buffers. Otherwise it would have been far simpler to simply named shared memory...

uecker replied:
Ok, I cherry-picked this to master with minor changes.

uecker commented on src/misc/io.c:
```diff
@@ -87,6 +89,49 @@ void io_register_output(const char* name)
+		if (0 == strcmp(name, iop->name)) {
+			tmp = iop->prev;
+			XFREE(iop->name);
+			xfree(iop); // Note the lowercase, GODDAMMIT!
```
iop has an indeterminate value after freeing, so strictly speaking you can not compare it.

uecker replied:
Nevermind, I just rewrote the list processing in idiomatic C to make it simpler. Only slightly tested.

uecker replied:
The value of the pointer itself becomes indeterminate after free. Actually, not only this pointer but all pointers which point to the object you have freed.

Takishima replied:
But the `free` function does nothing to the value of the pointer passed as an argument, or am I missing something ?

The call to the `XFREE()`only zeros-out the `iop->name` pointer and not `iop`itself. Then the call to `xfree(iop)` releases the memory located at the location pointed to by `iop` and nothing else. Agreed, the location pointed to by `iop`is now invalid but `iop` itself still exists as a variable holding an address. The comparison `iofiles == iop` should therefore not pose a problem, since I am only looking at the value of the pointer themselves, which are still the same as before the two calls to `xfree` as neither `iop` nor `iofiles` have been modified until this point.

When developing this code, I ran some extensive tests using `valgrind` and did not detect any memory leaks nor accesses to invalid memory.

uecker replied:
No, the value of the pointer is then indeterminate. A compiler may use this for optimization and make the comparison be always false.

Takishima commented:
I just did a new rebase to take the latest commits on the master branch into account.

uecker commented on src/fcopy.c:
```diff
@@ -0,0 +1,60 @@
+/* Copyright 2018. Damien Nguyen
```
Hi Damien,
would changing 'copy' to also work for simple copies also be OK for your use case instead of adding a new command? (I already have such a patch locally)?

Takishima replied:
Yes that would work. If you prefer not to merge these changes, I could add this functionality outside of BART when I embed it.

uecker replied:
Ok, but I then also add the necessary change to 'copy'. It is useful functionality, but I rather having another copy command.

uecker commented on utests/test_splines.c:
```diff
@@ -191,8 +191,8 @@ static bool test_bspline_zero(void)
 
-		double k2[9];
-		double c2[6];
+		double k2[9] = {0., 0., 0., 0., 0., 0., 0., 0., 0.};
```
I have to look at the underlying issue. Maybe something is broken in the spline code...

Takishima replied:
That error was really weird, it only happened when using `clang` to compile BART.

uecker replied:
Similar to the other one, but that turned out to be a bug in BART.

Takishima commented:
I've rebased the remaining changes onto the current master and removed the `fcopy` commit from the changes.

uecker commented:
Okay, I merged most of it in some way. Thank you for your work!

For the in memory function I used a different solution. Let' me no whether this works for you.

I also edited the python module without testing, I hope nothing broke...

Finally, I lost the CMake integration testing a while ago (April 29 "update compiler on travis"). As you know CMake much better than me: If you have time, I would appreciate if you could help restore this...

Takishima commented:
I will take a look at the CMake integration testing over the next few weeks.

uecker commented:
Thank you! We had it already, so maybe only a minor change is needed to make it work again.
